src/model.rs

```diff
@@ -896,48 +902,63 @@ impl Collections {
                 let by_frequency: HashMap<Idx<StopArea>, usize> = group_by_frequencies(stop_areas);
                 let most_frequent_stop_areas = find_indexes_with_max_frequency(by_frequency);
                 let biggest_stop_areas =
                     find_biggest_stop_area(most_frequent_stop_areas, collections);
                 find_first_by_alphabetical_order(biggest_stop_areas)
             }
 
-            let origin_stop_area =
-                find_stop_area_for(collections, route_id, |vj| vj.stop_times[0].stop_point_idx);
-            let destination_stop_area = find_stop_area_for(collections, route_id, |vj| {
-                vj.stop_times[vj.stop_times.len() - 1].stop_point_idx
+            let vehicle_journey_idxs = routes_to_vehicle_journeys
+                .get_corresponding_forward(&IdxSet::from_iter(std::iter::once(route_idx)));
+
+            let origin_stop_area = find_stop_area_for(collections, &vehicle_journey_idxs, |vj| {
+                vj.stop_times[0].stop_point_idx
             });
+            let destination_stop_area =
+                find_stop_area_for(collections, &vehicle_journey_idxs, |vj| {
+                    vj.stop_times[vj.stop_times.len() - 1].stop_point_idx
+                });
 
             if let (Some(origin_stop_area), Some(destination_stop_area)) =
                 (origin_stop_area, destination_stop_area)
             {
                 Ok((origin_stop_area, destination_stop_area))
             } else {
-                bail!("Failed to generate a `name` for route {}", route_id)
+                bail!(
+                    "Failed to generate a `name` for route {}",
+                    &collections.routes[route_idx].id
+                )
             }
         }
 
-        let mut routes = self.routes.take();
-        for mut route in &mut routes {
+        let mut route_names: HashMap<Idx<Route>, String> = HashMap::new();
+        let mut route_destination_ids: HashMap<Idx<Route>, Option<String>> = HashMap::new();
+        for (route_idx, route) in &self.routes {
             if route.name.is_empty() {
                 let (origin, destination) = skip_error_and_log!(
-                    find_best_origin_destination(&route.id, &self),
+                    find_best_origin_destination(route_idx, &self, routes_to_vehicle_journeys,),
                     LogLevel::Warn
                 );
-                route.destination_id = Some(destination.id.clone());
-                route.name = format!("{} - {}", origin.name, destination.name);
+                let route_name = format!("{} - {}", origin.name, destination.name);
+                route_names.insert(route_idx, route_name);
+                route_destination_ids.insert(route_idx, Some(destination.id.clone()));
             }
         }
-        self.routes = CollectionWithId::new(routes).unwrap();
+        for (route_idx, route_name) in route_names {
```

Comment:
Original code was already looping twice on all routes (worst case all the time) because of
```rust
for mut route in &mut routes {
  ...
}
```
and
```rust
self.routes = CollectionWithId::new(routes).unwrap();
```
On top of that, in term of memory, `CollectionWithId::new()` is recreating a `HashMap` with all the routes (so always worst case too).

In the new code, there is the same number of operations in the worst case (creation of a `HashMap` and 2 complete iterations) + an extra iteration... but only in the worst case. In the best case, no route is modified, and therefore the `HashMap` will be empty in memory and the second iteration and third iteration are not gonna happen (although the iterators are still gonna be created).

Also, 1, 2 or 3 iterations on the Routes is O(3N) which is equivalent to O(N) in complexity. Or another way to say this

```
for mut pet in &mut pets {
  pet.name = "foo"
  pet.animal = "bar"
}
```
is probably equivalent to
```
for mut pet in &mut pets {
  pet.name = "foo"
}
for mut pet in &mut pets {
  pet.animal = "bar"
}
```
(although, in Rust, the second solution might be slightly less efficient because there is the creation of 2 iterators instead of one).

So the question is : "are the creation of iterators more costly than the possibility to have reduced iterations on the second and third iterations and a reduced `HashMap`".

Reply:
As discussed, there is no `iter_mut` on CollectionWithId (using it was my suggestion).
And adding it is a nice idea but more heavy work, so it won't be done shortly if ever.

Given that, the pattern used looks good to me (and improvement on the map used is fine).